packages/relay/src/lib/subscriptionController.ts

```diff
@@ -0,0 +1,134 @@
+/*-
+ *
+ * Hedera JSON RPC Relay
+ *
+ * Copyright (C) 2023 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+// import { Poller } from './poller';
+import { Logger } from 'pino';
+import LRU from "lru-cache";
+import constants from "./constants";
+import crypto from "crypto";
+
+export interface Subscriber {
+    connection: any,
+    subscriptionId: string
+}
+
+const LOGGER_PREFIX = 'Subscriptions:';
+
+export class SubscriptionController {
+    // private poller: Poller;
+    private logger: Logger;
+    private subscriptions: {[key: string]: Subscriber[]};
+    private cache;
+
+    // constructor(poller: Poller, logger: Logger) {
+    constructor(logger: Logger) {
+        // this.poller = poller;
+        this.logger = logger;
+        this.subscriptions = {};
+        this.cache = new LRU({ max: constants.CACHE_MAX, ttl: 2000 });
+    }
+
+    createHash(data) {
+        return crypto.createHash('sha3').update(data.toString()).digest('hex');
+    }
+
+    // Generates a random 16 byte hex string
+    generateId() {
+        return "0x" + crypto.randomBytes(16).toString('hex');
+    }
+
+    subscribe(connection, event: string, filters?: {}) {
+
+        let tag: any = {event};
+        if (filters && Object.keys(filters).length) {
+            tag.filters = filters;
+        }
+
+        tag = JSON.stringify(tag);
+
+        if (!this.subscriptions[tag]) {
+            this.subscriptions[tag] = [];
+        }
+
+        const subId = this.generateId();
+
+        this.logger.info(`${LOGGER_PREFIX} New subscription ${subId}, listening for ${tag}`);
+
+        this.subscriptions[tag].push({
+            subscriptionId: subId,
+            connection
+        });
+
+        // this.poller.add(tag, this.notifySubscribers.bind(this, tag));
+        return subId;
+    }
+
+    unsubscribe(connection, subId?: string) {
+        const {id} = connection;
+
+        if (subId) {
+            this.logger.info(`${LOGGER_PREFIX} Unsubscribing connection ${id} from subscription ${subId}`);
+        }
+        else {
+            this.logger.info(`${LOGGER_PREFIX} Unsubscribing all instances of connection ${id}`);
+        }
+
+        for (const [tag, subs] of Object.entries(this.subscriptions)) {
```

Comment:
Not an issue at this scale but would it be an option to have a reverse map i.e. `subId` -> `tag` so that future traversals only need to cover 1 list of connections?

Reply:
That would mean that we have to maintain 2 maps, one for `subId -> tag` and one for `tag -> subId`, which would mean that we keep twice as much data in memory.

Having only 1 map means that we have to loop either in `unsubscribe` or in `notifySubscribers`. And since `notifySubscribers` will be called every time when there is new data, I think looping in `unsubscribe` is the better option.